include/multigrid.h

```diff
@@ -405,22 +404,59 @@ namespace quda {
     void saveVectors(const std::vector<ColorSpinorField *> &B) const;
 
     /**
-       @brief Generate the null-space vectors
+       @brief Initialize a set of vectors to random noise
+       @param B set of vectors
+    */
+    void initializeRandomVectors(const std::vector<ColorSpinorField*> &B) const;
+
+    /**
+       @brief Wrapping function that manages logic for constructing near-null vectors
+       @param refresh whether or not we're only refreshing near null vectors
+    */
+    void constructNearNulls(bool refresh = false);
+
+    /**
+       @brief Generate the null-space vectors via inverse iterations
        @param B Generated null-space vectors
-       @param refresh Whether we refreshing pre-exising vectors or starting afresh
+       @param iterations if non-zero, override the max number of iterations
     */
-    void generateNullVectors(std::vector<ColorSpinorField*> &B, bool refresh=false);
+    void generateInverseIterations(std::vector<ColorSpinorField*> &B, int iterations = 0);
+
+    /**
+       @brief Generate the null-space vectors via a Chebyshev filter; this approach is
+              based on arXiv:2103.05034, P. Boyle and A. Yamaguchi.
+       @param B Generated null-space vectors
+    */
+    void generateChebyshevFilter(std::vector<ColorSpinorField*> &B);
 
     /**
        @brief Generate lowest eigenvectors
+       @param B Generated null-space vectors
+       @param post_restrict whether or not we're generating extra eigenvectors after restricting
+              some fine eigenvectors; if true we override the requested n_conv in the multigrid
+              eigenvalue struct
     */
-    void generateEigenVectors();
+    void generateEigenvectors(std::vector<ColorSpinorField*> &B, bool post_restrict = false);
+
+    /**
+       @brief Generate near-null vectors via restricting finer near-nulls, generating extras if need be
+       @param refresh Whether or not we're only refreshing extra near nulls
+    */
+    void generateRestrictedVectors(bool refresh = false);
 
     /**
        @brief Build free-field null-space vectors
-       @param B Free-field null-space vectors
+       @param B Generated null-space vectors
+    */
+    void generateFreeVectors(std::vector<ColorSpinorField*> &B);
+
+    /**
+      @brief Orthonormalize a vector of ColorSpinorField, erroring out if
+             the fields are not sufficiently linearly independent
+      @param vecs vector of ColorSpinorFields to normalize
+      @param count number of near-null vectors to orthonormalize, default all
     */
-    void buildFreeVectors(std::vector<ColorSpinorField*> &B);
+    void orthonormalize_vectors(const std::vector<ColorSpinorField*>& vecs, int count = -1) const;
```

Comment:
Probably should keep consistency with the functions declared above here and make this `orthonormalizeVectors` ?

Reply:
Can do, or I can rip the band-aid off and switch more/all functions over to underscores. Thoughts?

Reply:
Let’s keep the bandaid here for now to keep things simple.

Reply:
https://github.com/lattice/quda/commit/5ac5288cf79843ba7c64c7cd48859130c48fc6e6